wfe/wfe.go

```diff
@@ -595,12 +635,16 @@ func (wfe *WebFrontEndImpl) verifyJWS(
 		Path:   request.RequestURI,
 	}
 	if expectedURL.String() != headerURL {
-		return nil, "", nil, acme.MalformedProblem(fmt.Sprintf(
+		return nil, acme.MalformedProblem(fmt.Sprintf(
 			"JWS header parameter 'url' incorrect. Expected %q, got %q",
 			expectedURL.String(), headerURL))
 	}
 
-	return payload, headerURL, pubKey, nil
+	return &authenticatedPOST{
+		postAsGet: payload == nil,
```

Comment:
I think this is incorrect. Shouldn't it be:

```
string(payload) == ""
```

Reply:
I don't think so. When I construct a JWS with Go-Jose for the body `[]byte{}` it ends up giving back a `nil` from `jws.Verify(pubKey`) on the server side.

It's not clear how I can produce a JWS with the body `""` since the Go semantics make this into `[]byte{}` and the server reads a `nil` payload from the produced JWS in that case.

Reply:
Talked out of band: I understand now and you're correct :+1:

Reply:
Fixed in https://github.com/letsencrypt/pebble/pull/162/commits/011b1270ef6686a3a544dbc622a6c2ff94f54be5

Reply:
My recommendation here was actually also incorrect. We want to accept `{"payload": ""}` and reject `{"payload": null}`. Go parses the latter into a `[]byte` by assigning `nil` to the slice, but if you convert a `nil` byte slice to a `string`, you actually get an empty string. Here's the most precise check:

`len(payload) == 0 && payload != nil`

Here's an example that demonstrates: https://play.golang.org/p/-1JwGwfmo1X

Reply:
IMO this confusion is good reason to think this proposal should be changed. I started a discussion on-list: https://mailarchive.ietf.org/arch/msg/acme/HWXOB39xv0TqW6tbILlBnCgMFCQ

Reply:
It looks like consensus is building towards this being the least-worst path forward despite the unpleasantness of the empty body check.

> Here's the most precise check:
> len(payload) == 0 && payload != nil

Implemented in this branch in 7da91f1